rdmd.d

```diff
@@ -674,42 +738,211 @@ private string[string] getDependencies(string rootModule, string workDir,
+        if (depsExitCode)
+        {
+            stderr.writefln("Failed: %(%s %)", depsGetter);
+            collectException(std.file.remove(depsFilename));
+            exit(depsExitCode);
+        }
+
+        if (!dryRun) deps = readDepsFile();
+    }
+
+    // Add exe as the root of the dependency graph.
+    auto exeNode = DepNode.other(exe);
+    deps[exeNode] = HashSet!DepNode();
+    deps.lookupOrInit(DepNode.source(rootModule)).insert(exeNode);
+        /* Reading the deps file above did make an entry for the root module,
+        but it may have used a different path to the same file.
+        This is why lookupOrInit is here.
+        Ideally, there should be only one entry per file. But one extra
+        entry isn't so bad, I suppose. */
+
+    // All dependencies specified through --extra-file
+    foreach (immutable f; extraFiles)
+    {
+        DepNode extraNode;
+        switch (f.extension)
+        {
+            case ".d": case ".di": extraNode = DepNode.source(f); break;
+            case objExt: extraNode = DepNode.object(f); break;
+            default: assert(false);
+        }
+        deps.lookupOrInit(extraNode).insert(exeNode);
+    }
+
+    // Need to add void main(){}?
+    if (addStubMain)
+    {
+        auto stubMain = buildPath(myOwnTmpDir, "stubmain.d");
+        std.file.write(stubMain, "void main(){}");
+        deps.lookupOrInit(DepNode.source(stubMain)).insert(exeNode);
+    }
+
+    // add compiler binary and config file as dependencies to all source files
+
+    void addDepenencyToAllSourceFiles(DepNode dependency)
+    {
+        deps.lookupOrInit(dependency);
+        foreach (dependent; deps.byKey)
+        {
+            if (dependent.type == DepNode.Type.source)
+                deps[dependency].insert(dependent);
+        }
+    }
+
+    // binary
+    immutable binaryNode = DepNode.other(which(compiler));
+    deps.lookupOrInit(binaryNode);
+    addDepenencyToAllSourceFiles(binaryNode);
+
+    // config file
+    {
+        immutable cmd = [compiler, "--help"];
+        auto pipes = pipeProcess(cmd);
+
+        auto pattern = regex(`^Config file: (.*)$`);
+        foreach (line; pipes.stdout.byLineCopy)
+        {
+            auto m = line.matchFirst(pattern);
+            if (m)
+            {
+                immutable confFile = m[1].strip;
+                // The config file is special: if missing, that's fine too. So
+                // add it as a dependency only if it actually exists.
+                if (confFile.exists)
+                {
+                    addDepenencyToAllSourceFiles(DepNode.other(m[1]));
+                }
+                break;
             }
-            // Fall through to rebuilding the deps file
+        }
+
+        if (int status = wait(pipes.pid))
+        {
+            stderr.writefln("Failed: %(%s %)", cmd);
+            exit(status);
         }
     }
 
-    immutable rootDir = dirName(rootModule);
+    return deps;
+}
```

Comment:
At 220 lines, this seems to be quite a long function. I think it should be broken into smaller ones.